### ISSUE TITLE
Difficulty converging power flow case

### ISSUE THREAD

**Issue by iagochavarry** [lanl-ansi/PowerModels.jl#793]
I'm having trouble converging a medium-large case (7k buses) within real operating conditions. 

With a flat start, the `run_ac_pf` arrives at a locally infeasible point. Using external commercial software, we were able to get a converged operating point. We used it as the initial point for the `run_ac_pf` and the model successfully converged.

Since we are facing an NL problem, converging the problem might be a difficult task. Do you have any suggestions on how we can treat this issue? We thought about using relaxations for initial values, adding slack variables to help convergency, changing to ACR, but so far nothing really helped.

The network file used with the converged point is linked [here](https://drive.google.com/file/d/1QSJb6QspFIL5PCDM7NG0IqyVrmRQBt8G/view?usp=sharing)

```julia
using PowerModels, Ipopt, JuMP, JSON

# NLP solver
const ipopt = optimizer_with_attributes(Ipopt.Optimizer,
                                        MOI.Silent() => false,
                                        "tol" => 0.1,
                                        "max_iter" => 200)


# parsing functions
read_json(file::String) = JSON.parse(String(read(file)))

# files
network = read_json("network.json") # dictionary of not converged file

# running without converged warm start
result = run_pf(network, ACPPowerModel, ipopt)

# running with converged warm start
set_ac_pf_start_values!(network)
result = run_pf(network, ACPPowerModel, ipopt)
```

**Comment by iagochavarry** [comment]
Thanks, @ccoffrin. I'll try your suggestions. 

We were very thrilled when performing the data conversion. All smaller cases tests suggest that the conversions are adequate.

Just reinforcing that the network data has a feasible solution. Even because setting `set_ac_pf_start_values!(network)` as the initial condition converges the case (this initial condition is the point converged with the commercial software).

I figure that since we found an initial condition that successfully converges the case (commercial software condition), the issue of running `run_ac_pf` with flat start is typical NL problem convergence.

I'll try `run_ac_opf` to further investigate, but I don't quite understand how the result of the `opf` is related to the `pf`.

**Comment by ccoffrin** [comment]
Very interesting. Given these results could I am wondering if the given PF solution is low voltage solution and there is no feasible solution within the desired voltage bounds?

I would surely have a close look at the `vm` values in the data to see what is going on in there.

**Comment by iagochavarry** [comment]
Make sense. The voltage profile of the data is indeed strange (several low voltages). I will take a closer look at them. Thanks a lot for the help!

**Comment by iagochavarry** [comment]
@ccoffrin one last question. In your experience, do you think exploring relaxations or other approximations of the problem as initial values for the AC PF can help?

**Comment by ccoffrin** [comment]
My experience has been that warm starting PF with solutions from relaxations and approximations is generally not helpful, as apposed to running AC-PF directly.  But maybe that was due to a lack of identified cases where Ipopt would not converge on the AC problem.  If you were going to try, I would recommend trying the LPACC model.  It tends to be close to the AC model in a wide range of operating conditions.

**Comment by ccoffrin** [comment]
@iagochavarry is there something actionable in this issue or can we close it?

**Comment by iagochavarry** [comment]
@ccoffrin I suppose we can close it. 

Let me give you more information about the convergence issues we are facing before :)

We have noticed this convergence issue for several of our system network data. 

These systems with convergence issues are the ones representing a relatively detailed version of the Brazilian power system (6k ~10k buses, 8k ~13k transmission lines)

This network is very stressed, with a voltage profile ranging between 0.6 p.u. to 1.3 p.u. (although, the vast majority of them are between 0.8 and 1.15 p.u.).

As far as we can tell, the issue relies on the difficulty of finding a feasible solution by the non-linear solvers, **since when we pass the initial point as the exact convergence point, the algorithm finds the feasible solution.**

We have been measuring the "difficulty of convergence" introducing slack variables into the power balance constraints and penalizing them in the objective function. The distance between the feasible solution is something like the value of the objective function.

We noticed that when we introduce some control actions in the problems, such as constraining the reactive generation between the generator's bounds, constraining the voltage magnitude of PQ buses within the bounds, the slack variables tend to assume lower values.

In the Brazilian system, the convergence of the cases are often obtained through a controlled power flow with actions such as mentioned above, and other ones, such as (i) controlling bus voltage with automatic tap variation, automatic shunt variation; (ii) controlling power flow with automatic phase shifter variation; (iii) remote control with PQV buses. This actions are performed to accommodate the stressed network.

We think that the difficulty of finding a feasible solution for the Power Flow might be related with the absence of those controlled actions, performed in classic commercial software in Brasil. Because of that, we are implementing a power flow with those classic control actions and verifying if this helps in the convergence of the problem. 

Nevertheless, we are exploring other approaches to solve our problem... In our framework, we are interested in the active power flow in some key lines of the system. Therefore, we are considering mixing relaxations, approximations, and other things to relieve the strong non-linearity/non-convexity, while keeping the active power flow in those key lines as close as possible to the original values. 

With that being said, do know any literature that might be helpful in approximating/relaxing the problem, while keeping the active power flow in a subset of the system close to the full AC formulation value? And what do you think about adding the controllable actions into the power flow?

**Comment by ccoffrin** [comment]
Thank you for this report @iagochavarry! I am going to close this issue for lack of actionable changes to PowerModels at this time, but the issue you identified is quite interesting.

To answer your question about approximations. Some years ago I developed the LPAC approximation (which is available in PowerModels) to handle these sorts of situations. In my experience it works well and finds solutions that can be converted to AC-feasible easily.  It can also be used in mixed integer solvers, like Gurobi.  So I can recommend this approach for finding high quality approximate solutions.

In practice, however, I have found that NLP solvers are faster and more accurate than approximations and relaxations.  It has been very challenging to find a case where some variant of AC-PF or AC-OPF is not the best solution in terms of accuracy and runtime.  If you have identified such a situation, archiving these specific cases and laying out the characteristics and challenges in solving for the reach community could be a fantastic contribution. If you would like to go in that direction send me and e-mail to discuss the details.
